File: api/src/main/java/dev/obscuria/fragmentum/registry/Deferred.java
package dev.obscuria.fragmentum.registry;

import net.minecraft.core.Holder;
import org.apache.commons.lang3.NotImplementedException;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public class Deferred<T, V extends T> implements HolderProvider<T>, ValueProvider<V>
{
    public Deferred(Supplier<Holder<T>> supplier) {}

    @Override
    public Holder<T> holder()
    {
        throw new NotImplementedException();
    }

    @Override
    public V get()
    {
        throw new NotImplementedException();
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/registry/Deferred.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.registry

import net.minecraft.core.Holder

open class Deferred<T, V : T>(supplier: () -> Holder<T>) : HolderProvider<T>, ValueProvider<V> {

    override val holder: Holder<T> = error("API stub")
    override val value: V = error("API stub")
}